### PR TITLE
Fixed deserialization of variable-length data types

### LINKED ISSUE #100: JSON deserialization
Hi. I've noticed you added JSON support. mysql-binlog-connector-java@master is missing [a number of changes](https://github.com/shyiko/mysql-binlog-connector-java/compare/0.5.0...0.5.2) that were included in 0.5.2 that you might be interested in. This one in particular (https://github.com/shyiko/mysql-binlog-connector-java/commit/881e9c785ff71a9f14ea8e44d0e6a3bc4346b535). Take a look. It might save you from a couple of bug reports. Cheers!

### PR DESCRIPTION
**Pull request by b13f**
Made fix according https://github.com/percona/percona-server/blob/5.7/sql/json_binary.cc#L247

Fixes #100

### CODE REVIEW

**Comment by siddontang** [comment]
CI failed

**Comment by siddontang** [comment]
Btw, what is wrong with current implementation ?

**Comment by b13f** [comment]
Some strings were rising error at isDataShort https://github.com/siddontang/go-mysql/blob/master/replication/json_binary.go#L339 when n>1
I checked C source and find issues with that fixes in https://github.com/shyiko/mysql-binlog-connector-java/commit/1612ebe3c21a248586500ae4f545bf33fff1d080#diff-1b0d47a97940724d9e6df0cfa9f04405R940
after that fix all my strings pass deserialisation

**Comment by siddontang** [comment]
Thanks @b13f 

Can you add a test for this?

**Comment by b13f** [comment]
@siddontang done!

File: replication/row_event_test.go [hunk 1]
```diff
@@ -420,12 +421,15 @@ func (_ *testDecodeSuite) TestParseJson(c *C) {
...
...t fringilla. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; In ac dui quis mi consectetuer lacinia. Nam pretium turpis et arcu. Duis arcu tortor, suscipit eget, imperdiet nec, imperdiet iaculis, ipsum. Sed aliquam ultrices mauris. Integer ante arcu, accumsan a, consectetuer eget, posuere ut, mauris. Praesent adipiscing. Phasellus ullamcorper ipsum rutrum nunc. Nunc nonummy metus. Vestibulum volutpat pretium libero. Cras id dui. Aenean ut eros et nisl sagittis vestibulum. Nullam nulla eros, ultricies sit amet, nonummy id, imperdiet feugiat, pede. Sed lectus. Donec mollis hendrerit risus. Phasellus nec sem in justo pellentesque facilisis. Etiam imperdiet imperdiet orci. Nunc nec neque. Phasellus leo dolor, tempus non, auctor et, hendrerit quis, nisi. Curabitur ligula sapien, tincidunt non, euismod vitae, posuere imperdiet, leo. Maecenas malesuada. Praesent congue erat at massa. Sed cursus turpis vitae tortor. Donec posuere vulputate arcu. Phasellus accumsan cursus velit. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; Sed aliquam, nisi quis porttitor congue, elit erat euismod orci, ac\x80\x00\x00\x00e"),
```
Comment (siddontang) [inline replication/row_event_test.go]: I think we can use a longTbls to test the long JSON data individually.

Comment (b13f) [inline replication/row_event_test.go]: I should use variable longTbls and use it like `for _, tbl := range longTbls` ? Is it correct?

Comment (siddontang) [inline replication/row_event_test.go]: yep +1

Comment (b13f) [inline replication/row_event_test.go]: :white_check_mark:

**Pull request merged**